### DISCUSSION TITLE
Using <dir> in mensural notation

### DISCUSSION THREAD

**Discussion by craigsapp** [rism-digital/verovio#3616] [Q&A] [answered]
Can `<dir>` be used in mensural music?  And is it implemented in verovio?  I cannot get it to display:

Target rendering (using CMN):

<img width="228" alt="Screenshot 2024-03-05 at 18 42 45" src="https://github.com/rism-digital/verovio/assets/3487289/6cd4f8f5-3ac5-4151-976e-bd57f8d1127b">

<details>
<summary> Click to view MEI data for above example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt>
   </pubStmt>
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2024-03-05T18:38:13" version="4.2.0-dev-84f99b0-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
 </meiHead>
 <music decls="#work0_encoded">
  <body>
   <mdiv xml:id="m180eaeo">
    <score xml:id="s1c7i7tz">
     <scoreDef xml:id="s1kc2bbk" midi.bpm="400.000000">
      <staffGrp xml:id="s1hamfyp">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="C" line="3" />
        <meterSig xml:id="metersig-L3F1" count="2" unit="1" visible="false" />
        <mensur xml:id="mensur-L4F1" sign="C" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="1" oct="4" pname="c" accid.ges="n" />
         <note xml:id="note-L7F1" dur="1" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <dir xml:id="dir-L6F1" place="above" staff="1" tstamp="2.000000">
        <rend xml:id="r7mq8oa" fontstyle="normal">above</rend>
       </dir>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

When trying to do the same in mensural notation, the `<dir>` does not display:

<img width="165" alt="Screenshot 2024-03-05 at 18 43 19" src="https://github.com/rism-digital/verovio/assets/3487289/1f2f421d-e446-4049-af7d-6c601cf01ebd">

<details>
<summary> Click to view MEI data for above example. </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/5.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt>
   </pubStmt>
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2024-03-05T18:34:19" version="4.2.0-dev-84f99b0-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mh7oymp">
    <score xml:id="skc7ui7">
     <scoreDef xml:id="smi7weh" midi.bpm="280.000000">
      <staffGrp xml:id="s14sva9l">
       <staffDef xml:id="staffdef-L1F1" n="1" notationtype="mensural.white" lines="5">
        <clef xml:id="clef-L2F1" shape="C" line="3" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <staff xml:id="staff-L1F1" n="1">
       <layer xml:id="layer-L1F1N1" n="1">
        <note xml:id="note-L5F1" dur="semibrevis" oct="4" pname="c" />
        <note xml:id="note-L8F1" dur="semibrevis" oct="4" pname="d" />
        <barLine xml:id="barline-L9F1" form="single" />
       </layer>
       <dir xml:id="dir-L7F1" place="above" staff="1" startid="note-L8F1">
        <rend xml:id="r162h8aw" fontstyle="normal">above</rend>
       </dir>
     </staff>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

There is no measure in the mensural MEI, but the `<dir>` documentation says that `<dir>` is allowed in `<staff>`:

https://music-encoding.org/guidelines/v5/elements/dir.html

<img width="924" alt="Screenshot 2024-03-05 at 18 51 09" src="https://github.com/rism-digital/verovio/assets/3487289/bb339489-7b3b-41f5-a888-52cc47e485a6">

I also tried in illegal places such as inside of `<layer>` and `<section>`, but it still does not display.

I am using `@startid` in the above mensural example, but I have also tried `@tstamp` (which is less useful in mensural music).

**Comment by lpugin** [comment, marked as answer]
This is a recurrent issue. When discussing this with @pe-ro at MEC there was the idea to introduce the concept (element) of an anchor that would be allowed within `layer`. If I remember correctly, we discussed the idea for it to be a container for control events (e.g. `dir`) since with mensural (un-measured) notation, the content of `layer` can basically be the whole piece and having all the control events piled up at the end of the `staff` is not very practical.

The result would be something like:
```xml
<layer>
   <note/>
   <anchor>
      <dir>above</dir>
   </anchor>
   <note/>
   <anchor>
      <fermata/>
   </anchor>
</layer>
```

That would avoid allowing all control events within `layer`, which would be quite problematic. Obviously more a MEI question from that point of view.

**Comment by craigsapp** [comment]
It would be useful to localize `<dir>` near  notes in mensural music without measures like the example `<fermata>`.

**Comment by craigsapp** [comment]
How would the `<anchor>` system work with two endpoints that are typically specified with `@tstamp`/`@tstamp2` or `@startid`/`@endid`?

Maybe something like this:

```xml
<layer>
   <note xml:id="A"/>
   <anchor>
      <bracket endid="#B"/>
   </anchor>
   <note/>
   <anchor>
      <fermata/>
   </anchor>
   <note>/>
   <note/>
   <note/>
   <note xml:id="B">
</layer>
```

with `@tstamp` being implicitly derived from `<anchor>`.

And another possibility might be to have `<anchor>` as a child of `<note>`:

```xml
<layer>
   <note xml:id="A">
      <anchor>
         <bracket endid="#B"/>
      </anchor>
   </note>
   <note/>
   <anchor>
      <fermata/>
   </anchor>
   <note>/>
   <note/>
   <note/>
   <note xml:id="B">
</layer>
```

Where `bracket@startid="#A"` is implicitly derived from the parent of `<anchor>`.

Using `bracket@tstamp2` (which would be difficult to specify in mensural notation):

```xml
<layer>
   <note/>
   <anchor>
      <bracket tstamp2="???"/>
   </anchor>
   </note>
   <note/>
   <anchor>
      <fermata/>
   </anchor>
   <note>/>
   <note/>
   <note/>
   <note/> <!-- matching tstamp2 for of this note -->
</layer>
```

Or adding an ID for another `<anchor>` for an easier solution to avoid `@tstamp2` using another`<anchor>` that may otherwise be empty and only defining a control point ending:

```xml
<layer>
   <note xml:id="A">
   <anchor>
      <bracket endid="#B"/>
   </anchor>
   </note>
   <note/>
   <anchor>
      <fermata/>
   </anchor>
   <note>/>
   <note/>
   <note/>
   <note/> 
   <anchor id:xml="B" />
</layer>
```

This would be equivalent to `<bracket tstamp="???" tstamp2="???"/>` but be implemented with `@tstamp2` derived implicitly from`@endid` of the second `<anchor>` element.

Adding `<anchor>` as a child of `<note>` (or other control points) simplifies processing of the `<bracket>` rendering (or other control points that would like to know the element types that they are connecting like bracket, where you want to include the ending note within the bracket and not have the bracket end at the start position of the note).  In other words, `<anchor>` as a sibling of `<note>` would be like `@tstamp`/`@tstamp2`, and `<anchor>` as a child of `<note>` would be like `@startid`/`@endid`.
